astroid/nodes/node_ng.py

```diff
@@ -248,14 +249,17 @@ def parent_of(self, node):
                 return True
         return False
 
-    def statement(self):
+    def statement(self) -> Union["nodes.Statement", "nodes.Module"]:
         """The first parent node, including self, marked as statement node.
 
         :returns: The first parent statement.
         :rtype: NodeNG
         """
         if self.is_statement:
+            self = cast("nodes.Statement", self)
             return self
+        if not self.parent:
+            raise ParentMissingError(target=self)
```

Comment:
@cdce8p Pinging you again. Please consider unsent if you just haven't had time to look at this!

Reply:
> (Just passing a `future` argument to functions doesn't really work because we would then deprecate the future parameter afterwards.)

Isn't that how `SQLAlechemy` will do it, too? I would recommend to add it as argument to `statement()`. Passing it during initialization has some drawbacks which I'd like to avoid:
* Difficult to pass around to all nodes, could introduce issues with circular dependencies. Although we might be able to make it work in the end.
* What if we want to add a `future` argument to another function? We would break code that has already added the flag.

Reply:
https://docs.sqlalchemy.org/en/14/core/future.html
https://github.com/sqlalchemy/sqlalchemy/blob/8656050241766688fee8363b5bb950d3945756c1/lib/sqlalchemy/engine/create.py#L627-L632

Sorry, I should have linked the documentation.
Seems to me that they pass the argument to two of their base classes as a "flag" and then methods/classes get changed depending on that flag. However, I have never used their code before, so it might not be true. 

I'm against passing it as an argument to `statement()` because then we would need to deprecate the `future` parameter again at some point. That seems like double-work?

Reply:
> Sorry, I should have linked the documentation. Seems to me that they pass the argument to two of their base classes as a "flag" and then methods/classes get changed depending on that flag. However, I have never used their code before, so it might not be true.

Yes, they pass it to the base functions. However, as I explained earlier, that doesn't really work for us. Should we add `future2`, `future3`, ... for any subsequent change to other functions? `SQLAlchemy` is different as they have just one big change that they what to do, not many small once.

> I'm against passing it as an argument to `statement()` because then we would need to deprecate the `future` parameter again at some point. That seems like double-work?

We need to deprecate it either way. If you don't like that, the only other alternatives I can think of would be to not add a breaking change or create a new method with a different name that replaces `statement()`. Not sure these are any better, though.

--
It would be a bit easier if we were changing function arguments. Then we could use something similar to what I've done for the removal of `None` as default values and use `inspect.signature` among other things. Without any `arguments` that doesn't work unfortunately.
https://github.com/PyCQA/astroid/blob/cdf8a2a9813128db5732a792434b89967c6fb701/astroid/decorators.py#L155-L209

Reply:
> We need to deprecate it either way. If you don't like that, the only other alternatives I can think of would be to not add a breaking change or create a new method with a different name that replaces `statement()`. Not sure these are any better, though.

Would the `future` parameter be deprecated at the same time as we change the normal behaviour? I think I (wrongly) assumed that the flow would be:
1) Allow `future` parameter
2) Bump to 3.0 > change normal behaviour
3) Warn of deprecation for `future` parameter
4) Bump to 4.0 > disallow `future` parameter
 
Thinking about this you probably meant to do:
1) Allow `future` parameter
2) Bump to 3.0 > change normal behaviour and disallow `future` parameter

The second option seems fine to me!

Reply:
I believe `SQLAlchemy` uses the first flow, although they might not wait till the next major release to disallow it.
For us, the second option should work. It's just a matter of removing the `future=True` at that point.

Reply:
I'll create a commit that implements this!